FastSimulation/TrackingRecHitProducer/plugins/TrackingRecHitProducer.cc

```diff
@@ -0,0 +1,247 @@
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Framework/interface/Run.h"
+#include "FWCore/Framework/interface/IOVSyncValue.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+
+#include "FastSimulation/TrackingRecHitProducer/interface/TrackingRecHitAlgorithm.h"
+#include "FastSimulation/TrackingRecHitProducer/interface/TrackingRecHitAlgorithmFactory.h"
+#include "FastSimulation/TrackingRecHitProducer/interface/TrackingRecHitPipe.h"
+
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+
+#include "SimDataFormats/TrackingHit/interface/PSimHit.h"
+
+#include "DataFormats/Common/interface/DetSetNew.h"
+#include "DataFormats/Common/interface/DetSetVectorNew.h"
+
+#include "DataFormats/DetId/interface/DetId.h"
+#include "DataFormats/Common/interface/RefVector.h"
+
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/Records/interface/TrackerTopologyRcd.h"
+
+#include "FastSimulation/TrackingRecHitProducer/interface/TrackerDetIdSelector.h"
+
+#include "DataFormats/TrackerRecHit2D/interface/FastSingleTrackerRecHit.h"
+#include "DataFormats/TrackerRecHit2D/interface/FastTrackerRecHitCollection.h"
+
+#include <map>
+#include <memory>
+#include <vector>
+
+class TrackingRecHitProducer:
+    public edm::stream::EDProducer<>
+{
+    private:
+        edm::EDGetTokenT<std::vector<PSimHit>> _simHitToken;
+        std::vector<TrackingRecHitAlgorithm*> _recHitAlgorithms;
+        edm::IOVSyncValue _iovSyncValue;
+        std::map<unsigned int, TrackingRecHitPipe> _detIdPipes;
+        void setupDetIdPipes(const edm::EventSetup& eventSetup);
+
+    public:
+        TrackingRecHitProducer(const edm::ParameterSet& config);
+
+        virtual void beginRun(edm::Run const&, const edm::EventSetup& eventSetup);
+
+        virtual void beginStream(edm::StreamID id);
+
+        virtual void produce(edm::Event& event, const edm::EventSetup& eventSetup);
+
+        virtual void endStream();
+
+        virtual ~TrackingRecHitProducer();
+};
+
+
+TrackingRecHitProducer::TrackingRecHitProducer(const edm::ParameterSet& config)
+{
+    edm::ConsumesCollector consumeCollector = consumesCollector();
+    const std::vector<edm::ParameterSet>& pluginConfigs = config.getParameter<std::vector<edm::ParameterSet>>("plugins");
+
+    for (unsigned int iplugin = 0; iplugin<pluginConfigs.size(); ++iplugin)
+    {
+        const edm::ParameterSet& pluginConfig = pluginConfigs[iplugin];
+        const std::string pluginType = pluginConfig.getParameter<std::string>("type");
+        const std::string pluginName = pluginConfig.getParameter<std::string>("name");
+
+        TrackingRecHitAlgorithm* recHitAlgorithm = TrackingRecHitAlgorithmFactory::get()->tryToCreate(pluginType,pluginName,pluginConfig,consumeCollector);
```

Comment:
just added proper deletion of these plugins - however, this would not have lead to an increasing memleak